File: CloudFlow/FrontEnd-App/src/app/shared/model/content/metadata-definition.model.ts
import { JsonObject, JsonProperty } from "json2typescript";
import { ListMetadataItemModel } from "./list-metadata-item.model";
import { ArrayUtils } from "./../../utils/array-utils";
import { NomenclatorMetadataDefinitionValueSelectionFilterModel } from "./nomenclator-metadata-definition-value-selection-filter.model";

@JsonObject
export class MetadataDefinitionModel {
	
	public static readonly TYPE_TEXT: string = "TEXT";
	public static readonly TYPE_NUMERIC: string = "NUMERIC";
	public static readonly TYPE_AUTO_NUMBER: string = "AUTO_NUMBER";
	public static readonly TYPE_DATE: string = "DATE";
	public static readonly TYPE_DATE_TIME: string = "DATE_TIME";
	public static readonly TYPE_MONTH: string = "MONTH";
	public static readonly TYPE_LIST: string = "LIST";
	public static readonly TYPE_USER: string = "USER";
	public static readonly TYPE_TEXT_AREA: string = "TEXT_AREA";
	public static readonly TYPE_METADATA_COLLECTION: string = "METADATA_COLLECTION";
	public static readonly TYPE_NOMENCLATOR: string = "NOMENCLATOR";
	public static readonly TYPE_GROUP: string = "GROUP";
	public static readonly TYPE_DOCUMENT: string = "DOCUMENT";
	public static readonly TYPE_CALENDAR: string = "CALENDAR";
	public static readonly TYPE_PROJECT: string = "PROJECT";
	
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_USER_TITLE: string = "USER_TITLE";
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_USER_EMAIL: string = "USER_EMAIL";
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_NOMENCLATOR_ATTRIBUTE: string = "NOMENCLATOR_ATTRIBUTE";
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_METADATA_VALUE: string = "METADATA_VALUE";
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_DELETE_VALUE: string = "DELETE_VALUE";
	public static readonly TYPE_OF_AUTO_COMPLETE_WITH_METADATA_CUSTOM_CLASS: string = "CUSTOM_CLASS";

	@JsonProperty("id", Number)
	public id: number = null;

	@JsonProperty("name", String)
	public name: string = null;

	@JsonProperty("label", String)
	public label: string = null;

	@JsonProperty("mandatory", Boolean)
	public mandatory: boolean = false;

	@JsonProperty("restrictedOnEdit", Boolean)
	public restrictedOnEdit: boolean = false;

	@JsonProperty("mandatoryStates", String)
	public mandatoryStates: string = null;

	@JsonProperty("restrictedOnEditStates", String)
	public restrictedOnEditStates: string = null;

	@JsonProperty("invisible", Boolean)
	public invisible: boolean = false;

	@JsonProperty("invisibleInStates", String)
	public invisibleInStates: string = null;

	@JsonProperty("representative", Boolean)
	public representative: boolean = false;

	@JsonProperty("indexed", Boolean, true)
	public indexed: boolean = undefined;

	@JsonProperty("orderNumber", Number)
	public orderNumber: number = null;

	@JsonProperty("type", String)
	public type: string = null;

	@JsonProperty("defaultValue", String, true)
	public defaultValue: string = undefined;

	@JsonProperty("metadataNameForAutoCompleteWithMetadata", String, true)
	public metadataNameForAutoCompleteWithMetadata: string = null;

	@JsonProperty("typeOfAutoCompleteWithMetadata", String, true)
	public typeOfAutoCompleteWithMetadata: string = undefined;

	@JsonProperty("nomenclatorAttributeKeyForAutoCompleteWithMetadata", String, true)
	public nomenclatorAttributeKeyForAutoCompleteWithMetadata: string = null;

	@JsonProperty("classNameForAutoCompleteWithMetadata", String, true)
	public classNameForAutoCompleteWithMetadata: string = null;

	// Date Metadata
	@JsonProperty("autoCompleteWithCurrentDate", Boolean, true)
	public autoCompleteWithCurrentDate: boolean = undefined;

	// DateTime Metadata
	@JsonProperty("autoCompleteWithCurrentDateTime", Boolean, true)
	public autoCompleteWithCurrentDateTime: boolean = undefined;

	// AutoNumber Metadata
	@JsonProperty("prefix", String, true)
	public prefix: string = undefined;

	@JsonProperty("numberLength", Number, true)
	public numberLength: number = undefined;

	// List Metadata
	@JsonProperty("multipleSelection", Boolean, true)
	public multipleSelection: boolean = undefined;

	@JsonProperty("extendable", Boolean, true)
	public extendable: boolean = undefined;

	@JsonProperty("listItems", [ListMetadataItemModel], true)
	public listItems: ListMetadataItemModel[] = undefined;

	// User Metadata
	@JsonProperty("onlyUsersFromGroup", Boolean, true)
	public onlyUsersFromGroup: boolean = undefined;

	@JsonProperty("idOfGroupOfPermittedUsers", String, true)
	public idOfGroupOfPermittedUsers: string = undefined;
	
	@JsonProperty("autoCompleteWithCurrentUser", Boolean, true)
	public autoCompleteWithCurrentUser: boolean = undefined;

	@JsonProperty("autoCompleteWithCurrentUserStateCode", String, true)
	public autoCompleteWithCurrentUserStateCode: string = undefined;

	// Collection Metadata
	@JsonProperty("metadataDefinitions", [MetadataDefinitionModel], true)
	public metadataDefinitions: MetadataDefinitionModel[] = undefined;

	// Nomenclator
	@JsonProperty("nomenclatorId", Number, true)
	public nomenclatorId: number = undefined;

	@JsonProperty("valueSelectionFilters", [NomenclatorMetadataDefinitionValueSelectionFilterModel], true)
	public valueSelectionFilters: NomenclatorMetadataDefinitionValueSelectionFilterModel[] = undefined;

	// Month
	@JsonProperty("autoCompleteWithCurrentMonth", Boolean, true)
	public autoCompleteWithCurrentMonth: boolean = undefined;

	// Document
	@JsonProperty("metadataDocumentTypeId", Number, true)
	public metadataDocumentTypeId: number = undefined;

	@JsonProperty("multipleDocumentsSelection", Boolean, true)
	public multipleDocumentsSelection: boolean = undefined;
	
	// Project
	@JsonProperty("multipleProjectsSelection", Boolean, true)
	public multipleProjectsSelection: boolean = undefined;
	
	public clone(): MetadataDefinitionModel {
		let clone: MetadataDefinitionModel = new MetadataDefinitionModel;
		clone.id = this.id;
		clone.name = this.name;
		clone.label = this.label;		
		clone.mandatory = this.mandatory;
		clone.restrictedOnEdit = this.restrictedOnEdit;
		clone.mandatoryStates = this.mandatoryStates;
		clone.restrictedOnEditStates = this.restrictedOnEditStates;
		clone.invisible = this.invisible;
		clone.invisibleInStates = this.invisibleInStates;
		clone.representative = this.representative;
		clone.indexed = this.indexed;
		clone.orderNumber = this.orderNumber;
		clone.type = this.type;
		clone.defaultValue = this.defaultValue;
		clone.metadataNameForAutoCompleteWithMetadata = this.metadataNameForAutoCompleteWithMetadata;
		clone.typeOfAutoCompleteWithMetadata = this.typeOfAutoCompleteWithMetadata;
		clone.nomenclatorAttributeKeyForAutoCompleteWithMetadata = this.nomenclatorAttributeKeyForAutoCompleteWithMetadata;
		clone.classNameForAutoCompleteWithMetadata = this.classNameForAutoCompleteWithMetadata;
		clone.autoCompleteWithCurrentDate = this.autoCompleteWithCurrentDate;
		clone.autoCompleteWithCurrentDateTime = this.autoCompleteWithCurrentDateTime;
		clone.prefix = this.prefix;
		clone.numberLength = this.numberLength;
		clone.multipleSelection = this.multipleSelection;
		clone.extendable = this.extendable;		
		if (ArrayUtils.isNotEmpty(this.listItems)) {
			clone.listItems = [];
			this.listItems.forEach((lmiModel: ListMetadataItemModel) => {
				clone.listItems.push(lmiModel.clone());
			});
		}
		clone.onlyUsersFromGroup = this.onlyUsersFromGroup;
		clone.idOfGroupOfPermittedUsers = this.idOfGroupOfPermittedUsers;
		clone.autoCompleteWithCurrentUser = this.autoCompleteWithCurrentUser;
		clone.autoCompleteWithCurrentUserStateCode = this.autoCompleteWithCurrentUserStateCode;
		if (ArrayUtils.isNotEmpty(this.metadataDefinitions)) {
			clone.metadataDefinitions = [];
			this.metadataDefinitions.forEach((mdModel: MetadataDefinitionModel) => {
				clone.metadataDefinitions.push(mdModel.clone());
			});
		}
		clone.nomenclatorId = this.nomenclatorId;
		if (ArrayUtils.isNotEmpty(this.valueSelectionFilters)) {
			clone.valueSelectionFilters = [];
			this.valueSelectionFilters.forEach((filterModel: NomenclatorMetadataDefinitionValueSelectionFilterModel) => {
				clone.valueSelectionFilters.push(filterModel.clone());
			});
		}
		clone.autoCompleteWithCurrentMonth = this.autoCompleteWithCurrentMonth;
		clone.metadataDocumentTypeId = this.metadataDocumentTypeId;
		clone.multipleDocumentsSelection = this.multipleDocumentsSelection;
		clone.multipleProjectsSelection = this.multipleProjectsSelection;
		return clone;
	}
}

File: CloudFlow/src/ro/cloudSoft/cloudDoc/presentation/client/shared/model/content/MetadataDefinitionModel.java
package ro.cloudSoft.cloudDoc.presentation.client.shared.model.content;

import java.util.Arrays;
import java.util.List;

import com.extjs.gxt.ui.client.data.BaseModel;
import com.google.gwt.user.client.rpc.IsSerializable;

import ro.cloudSoft.cloudDoc.presentation.client.shared.model.ModelWithLabel;

public class MetadataDefinitionModel extends BaseModel implements ModelWithLabel, IsSerializable {
	
	public static final int LENGTH_NAME = 128;
	public static final int LENGTH_LABEL = 128;

	private static final long serialVersionUID = -1798047104878308777L;
	
	public static final String TYPE_TEXT = "TEXT";
	public static final String TYPE_NUMERIC = "NUMERIC";
	public static final String TYPE_AUTO_NUMBER = "AUTO_NUMBER";
	public static final String TYPE_DATE = "DATE";
	public static final String TYPE_DATE_TIME = "DATE_TIME";
	public static final String TYPE_MONTH = "MONTH";
	public static final String TYPE_LIST = "LIST";
	public static final String TYPE_USER = "USER";
	public static final String TYPE_TEXT_AREA = "TEXT_AREA";
	public static final String TYPE_METADATA_COLLECTION = "METADATA_COLLECTION";
	public static final String TYPE_NOMENCLATOR = "NOMENCLATOR";
	public static final String TYPE_GROUP = "GROUP";
	public static final String TYPE_DOCUMENT = "DOCUMENT";
	public static final String TYPE_CALENDAR = "CALENDAR";
	public static final String TYPE_PROJECT = "PROJECT";
	
	public static final List<String> TYPES = Arrays.asList(new String[] {
		TYPE_TEXT,
		TYPE_NUMERIC,
		TYPE_AUTO_NUMBER,
		TYPE_DATE,
		TYPE_DATE_TIME,
		TYPE_MONTH,
		TYPE_LIST,
		TYPE_USER,
		TYPE_TEXT_AREA,
		TYPE_METADATA_COLLECTION,
		TYPE_NOMENCLATOR,
		TYPE_GROUP,
		TYPE_DOCUMENT,
		TYPE_CALENDAR
	});
	
	public static final List<String> TYPES_METADATA_FROM_COLLECTION =
			Arrays.asList(new String[] {
		TYPE_TEXT,
		TYPE_NUMERIC,
		TYPE_AUTO_NUMBER,
		TYPE_DATE,
		TYPE_LIST
	});
	
	public static final List<String> SIMPLE_TYPES = Arrays.asList(new String[] {
			TYPE_TEXT,
			TYPE_NUMERIC,
			TYPE_TEXT_AREA,			
	});

	public static final String PROPERTY_ID = "id";
	public static final String PROPERTY_NAME = "name";
	public static final String PROPERTY_LABEL = "label";
	public static final String PROPERTY_MANDATORY = "mandatory";
	public static final String PROPERTY_RESTRICTED_ON_EDIT = "restrictedOnEdit";
	public static final String PROPERTY_MANDATORY_STATES = "mandatoryStates";
	public static final String PROPERTY_RESTRICTED_ON_EDIT_STATES = "restrictedOnEditStates";
	public static final String PROPERTY_INVISIBLE = "invisible";
	public static final String PROPERTY_INVISIBLE_IN_STATES = "invisibleInStates";
	public static final String PROPERTY_REPRESENTATIVE = "representative";
	public static final String PROPERTY_INDEXED = "indexed";
	public static final String PROPERTY_ORDER_NUMBER = "orderNumber";
	public static final String PROPERTY_TYPE = "type";
	public static final String PROPERTY_DEFAULT_VALUE = "defaultValue";
	
	private String metadataNameForAutoCompleteWithMetadata;
	private String typeOfAutoCompleteWithMetadata;
	private String nomenclatorAttributeKeyForAutoCompleteWithMetadata;
	private String classNameForAutoCompleteWithMetadata;
	
	public MetadataDefinitionModel() {}
	
	public void setId(Long id){
		set(PROPERTY_ID, id);
	}
	
	public Long getId(){
		return get(PROPERTY_ID); 
	}
	
	public void setName(String name){
		set(PROPERTY_NAME, name);
	}
	
	public String getName(){
		return get(PROPERTY_NAME); 
	}
	
	public void setLabel(String label){
		set(PROPERTY_LABEL, label);
	}
	
	public String getLabel(){
		return get(PROPERTY_LABEL); 
	}
	
	public void setMandatory(Boolean mandatory){
		set(PROPERTY_MANDATORY, mandatory);
	}
	
	public Boolean isMandatory(){
		return get(PROPERTY_MANDATORY); 
	}
	
	public void setMandatoryStates(String mandatoryStates ){
		set( PROPERTY_MANDATORY_STATES, mandatoryStates );
	}
	
	public String getMandatoryStates(){
		return get( PROPERTY_MANDATORY_STATES );
	}
	
	public void setRestrictedOnEdit(Boolean restrictedOnEdit){
		set(PROPERTY_RESTRICTED_ON_EDIT, restrictedOnEdit);
	}
	
	public Boolean isRestrictedOnEdit(){
		return get(PROPERTY_RESTRICTED_ON_EDIT); 
	}
	
	public void setRestrictedOnEditStates(String restrictedOnEditStates) {
		set(PROPERTY_RESTRICTED_ON_EDIT_STATES, restrictedOnEditStates);
	}
	public String getRestrictedOnEditStates() {
		return get(PROPERTY_RESTRICTED_ON_EDIT_STATES);
	}
	
	public void setInvisible(Boolean invisible){
		set(PROPERTY_INVISIBLE, invisible);
	}
	
	public Boolean isInvisible(){
		return get(PROPERTY_INVISIBLE); 
	}
	
	public void setInvisibleInStates(String invisibleInStates) {
		set(PROPERTY_INVISIBLE_IN_STATES, invisibleInStates);
	}
	public String getInvisibleInStates() {
		return get(PROPERTY_INVISIBLE_IN_STATES);
	}
	
	public void setRepresentative(Boolean representative){
		set(PROPERTY_REPRESENTATIVE, representative);
	}
	
	public Boolean isRepresentative(){
		return get(PROPERTY_REPRESENTATIVE); 
	}
	
	public void setIndexed(Boolean indexed){
		set(PROPERTY_INDEXED, indexed);
	}
	
	public Boolean isIndexed(){
		return get(PROPERTY_INDEXED); 
	}
	
	public void setOrderNumber(Integer orderNumber){
		set(PROPERTY_ORDER_NUMBER, orderNumber);
	}
	
	public Integer getOrderNumber(){
		return get(PROPERTY_ORDER_NUMBER);
	}
	
	public void setType(String type){
		set(PROPERTY_TYPE, type);
	}
	
	public String getType(){
		return get(PROPERTY_TYPE); 
	}
	
	public void setDefaultValue(String defaultValue) {
		set(PROPERTY_DEFAULT_VALUE, defaultValue);
	}
	
	public String getDefaultValue() {
		return get(PROPERTY_DEFAULT_VALUE);
	}

	public String getMetadataNameForAutoCompleteWithMetadata() {
		return metadataNameForAutoCompleteWithMetadata;
	}

	public void setMetadataNameForAutoCompleteWithMetadata(String metadataForAutoCompleteWithMetadata) {
		this.metadataNameForAutoCompleteWithMetadata = metadataForAutoCompleteWithMetadata;
	}

	public String getTypeOfAutoCompleteWithMetadata() {
		return typeOfAutoCompleteWithMetadata;
	}

	public void setTypeOfAutoCompleteWithMetadata(String typeOfAutoCompleteWithMetadata) {
		this.typeOfAutoCompleteWithMetadata = typeOfAutoCompleteWithMetadata;
	}

	public String getNomenclatorAttributeKeyForAutoCompleteWithMetadata() {
		return nomenclatorAttributeKeyForAutoCompleteWithMetadata;
	}
	
	public void setNomenclatorAttributeKeyForAutoCompleteWithMetadata(
			String nomenclatorAttributeKeyForAutoCompleteWithMetadata) {
		this.nomenclatorAttributeKeyForAutoCompleteWithMetadata = nomenclatorAttributeKeyForAutoCompleteWithMetadata;
	}
	
	public String getClassNameForAutoCompleteWithMetadata() {
		return classNameForAutoCompleteWithMetadata;
	}
	
	public void setClassNameForAutoCompleteWithMetadata(String classNameForAutoCompleteWithMetadata) {
		this.classNameForAutoCompleteWithMetadata = classNameForAutoCompleteWithMetadata;
	}
}